aten/src/ATen/native/cuda/FractionalMaxPool2d.cu

```diff
@@ -0,0 +1,367 @@
+#include "ATen/ATen.h"
+#include "ATen/AccumulateType.h"
+#include "ATen/cuda/CUDAApplyUtils.cuh"
+#include "ATen/cuda/CUDAContext.h"
+#include "ATen/cuda/detail/IndexUtils.cuh"
+#include "ATen/cuda/detail/TensorInfo.cuh"
+#include "ATen/cuda/detail/KernelUtils.h"
+#include "ATen/NativeFunctions.h"
+#include "ATen/TensorUtils.h"
+#include "ATen/Utils.h"
+#include "c10/util/Exception.h"
+
+#include <algorithm>
+#include <cfloat>
+#include <cmath>
+
+namespace at {
+namespace native {
+
+using namespace at::cuda::detail;
+
+namespace {
+
+template <typename scalar_t, typename accscalar_t>
+__device__ inline int get_interval(accscalar_t sample,
+  int index, int inputSize, int outputSize, int poolSize) {
+  accscalar_t alpha = (accscalar_t)(inputSize - poolSize) / (accscalar_t) (outputSize - 1);
+  if (index == outputSize - 1) {
+    return inputSize - poolSize;
+  } else {
+    return (int) ((index + sample) * alpha) - (int) (sample * alpha);
+  }
+}
+
+template <typename scalar_t, typename index_t>
+__device__ inline scalar_t* get_ref_by_coord(
```

Comment:
This file looks quite a bit different than `aten/src/THCUNN/generic/SpatialFractionalMaxPooling.cu` and I am wondering how you went about doing the port. Some notable differences I see:

1. The tensor structs no longer statically know their dimensionality, and use TensorInfo now. There must be some back story to why you chose to use TensorInfo here (I didn't follow the intermediate commits, so I can only guess), but unfortunately it's the wrong choice here. TensorInfo is written in such a way that it can work with arbitrary dimension tensors, and that comes at the cost of efficiency and ease-of-use. We should use the right tool for the right job, and in this case, because this is specifically a 2D kernel, we know the dimensionality statically, and thus, PackedTensorAccessor is the right tool for the job.

Reply:
Back story: there is an optional batch dimension for both input and output tensor, so input&output can be 3D or 4D. The DeviceTensor has an upcast method that adds the batch dimension if it is missing, coercing everything to be 4D. I did not find the upcast method for PackedTensorAccessor. Instead I found some other PR that used TensorInfo in such cases.

Perhaps I should use "resize_" as a replacement of "upcast"?

Reply:
Yeah, so for upcasting, there's a pretty simple trick: do an unsqueeze(0) on the tensor before grabbing the accessor with the correct dimensionality.